File: Remote/Source/Remote Elements/Views/ModeSelectionView.m
//
// ModeSelectionView.m
// Remote
//
// Created by Jason Cardwell on 3/20/13.
// Copyright 2013 Moondeer Studios. All rights reserved.
//
#import "RemoteElementView_Private.h"
#import "ButtonGroup.h"
#import "Button.h"
#import "Remote.h"
#import "RemoteController.h"

static int ddLogLevel   = DefaultDDLogLevel;
static int msLogContext = (LOG_CONTEXT_REMOTE | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);

@interface ModeSelectionView ()
@property (nonatomic, weak) ButtonView * selectedButton;
@end

@implementation ModeSelectionView

////////////////////////////////////////////////////////////////////////////////
#pragma mark - ButtonGroupView overrides
////////////////////////////////////////////////////////////////////////////////

- (void)addSubelementView:(ButtonView *)view {
  [super addSubelementView:view];

  if (!_selectedButton) [self selectButton:view];

  __weak ModeSelectionView * weakself = self;
  __weak ButtonView        * weakview = view;
  view.tapAction = ^{ [weakself handleSelection:weakview]; };

}

////////////////////////////////////////////////////////////////////////////////
#pragma mark - Selection Handling
////////////////////////////////////////////////////////////////////////////////

- (void)selectButton:(ButtonView *)newSelection {
  if (_selectedButton != newSelection && StringIsNotEmpty(newSelection.key)) {

    if (_selectedButton) _selectedButton.model.selected = NO;
    newSelection.model.selected = YES;
    _selectedButton = newSelection;

    RemoteController * controller = [RemoteController remoteController:self.model.managedObjectContext];
    controller.currentRemote.currentMode = newSelection.key;

  }
}

/// Button action attached to the view's button's as they are added as subviews. This method
/// updates the value of `selectedButton` with the `ButtonView` that invoked the method.
/// @param sender The `ButtonView` that has been touched.
- (void)handleSelection:(ButtonView *)sender {
  if (_selectedButton != sender) [self selectButton:sender];
  if (self.model.autohide) [self performSelector:@selector(tuck) withObject:nil afterDelay:1.0];
}

- (void)drawBackdropInContext:(CGContextRef)ctx inRect:(CGRect)rect {
  REPanelLocation panelLocation = self.model.panelLocation;

  CGContextClearRect(ctx, self.bounds);

  NSUInteger roundedCorners = (panelLocation == REPanelLocationRight
                               ? UIRectCornerTopLeft | UIRectCornerBottomLeft
                               : (panelLocation == REPanelLocationLeft
                                  ? UIRectCornerTopRight | UIRectCornerBottomRight
                                  : 0));

  UIBezierPath * bezierPath = [UIBezierPath bezierPathWithRoundedRect:self.bounds
                                                    byRoundingCorners:roundedCorners
                                                          cornerRadii:CGSizeMake(15, 15)];

  self.borderPath = bezierPath;

  [defaultBGColor() setFill];

  [bezierPath fillWithBlendMode:kCGBlendModeNormal alpha:0.9];

  CGRect  insetRect = CGRectInset(self.bounds, 0, 3);
  CGFloat tx        = (panelLocation == REPanelLocationRight ? 3 : -3);

  insetRect = CGRectApplyAffineTransform(insetRect, CGAffineTransformMakeTranslation(tx, 0));

  bezierPath = [UIBezierPath bezierPathWithRoundedRect:insetRect
                                     byRoundingCorners:roundedCorners
                                           cornerRadii:CGSizeMake(12, 12)];
  bezierPath.lineWidth = 2.0;
  [bezierPath strokeWithBlendMode:kCGBlendModeClear alpha:1.0];
}

@end

File: Remote/Source/Remote Elements/Views/ModeSelectionView.swift
//
//  ModeSelectionView.swift
//  Remote
//
//  Created by Jason Cardwell on 11/07/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import MoonKit

class ModeSelectionView: ButtonGroupView {

  private weak var selectedButton: ButtonView!
  
  /**
  addSubelementView:

  :param: view RemoteElementView
  */
  override func addSubelementView(view: RemoteElementView) {
    super.addSubelementView(view)
    if let buttonView = view as? ButtonView {
      if selectedButton == nil { selectButton(buttonView) }
      buttonView.tapAction = {self.handleSelection(buttonView)}
    }
  }

  /**
  selectButton:

  :param: newSelection ButtonView
  */
  func selectButton(newSelection: ButtonView) {
    if selectedButton != newSelection && newSelection.model.key != nil && !newSelection.model.key!.isEmpty {
      (selectedButton?.model as Button).selected = false
      (newSelection.model as Button).selected = true
      selectedButton = newSelection
      RemoteController(model.managedObjectContext).currentRemote?.currentMode = newSelection.model.key!
    }
  }

  /**
  handleSelection:

  :param: sender ButtonView
  */
  func handleSelection(sender: ButtonView) {
    if selectedButton != sender { selectButton(sender) }
    if (model as ButtonGroup).autohide { MSDelayedRunOnMain(1){self.tuck()} }
  }

  /**
  drawBackdropInContext:inRect:

  :param: ctx CGContextRef
  :param: inRect CGRect
  */
  override func drawBackdropInContext(ctx: CGContextRef, inRect: CGRect) {
    let panelLocation = (model as ButtonGroup).panelLocation
    CGContextClearRect(ctx, bounds)
    var roundedCorners: UIRectCorner
    switch panelLocation {
      case .Right: roundedCorners = UIRectCorner.TopLeft | UIRectCorner.BottomLeft
      case .Left: roundedCorners = UIRectCorner.TopRight | UIRectCorner.BottomRight
      default: roundedCorners = UIRectCorner(0)
    }
    var path = UIBezierPath(roundedRect: bounds, byRoundingCorners: roundedCorners, cornerRadii: CGSize(square: 15))
    borderPath = path
    UIColor.darkTextColor().setFill()
    path.fillWithBlendMode(kCGBlendModeNormal, alpha: 0.9)
    let dx: CGFloat = panelLocation == .Right ? 3 : -3
    let insetRect = bounds.rectByInsetting(dx: 0, dy: 3).rectByApplyingTransform(CGAffineTransformMakeTranslation(dx, 0))
    path = UIBezierPath(roundedRect: insetRect, byRoundingCorners: roundedCorners, cornerRadii: CGSize(square: 12))
    path.lineWidth = 2.0
    path.strokeWithBlendMode(kCGBlendModeClear, alpha: 1.0)
  }

}
